reconcile/cluster_auth_rhidp/integration.py

```diff
@@ -0,0 +1,248 @@
+from reconcile.utils.ocm.clusters import discover_clusters_for_organizations
+from reconcile.utils.ocm.label_sources import ClusterRef, LabelState
+from reconcile.utils.ocm.labels import add_label, delete_label, update_label
+from reconcile.utils.ocm_base_client import (
+    OCMAPIClientConfigurationProtocol,
+    OCMBaseClient,
+    init_ocm_base_client,
+)
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import SecretReaderBase
+from reconcile.utils.semver_helper import make_semver
+
+QONTRACT_INTEGRATION = "cluster-auth-rhidp"
+QONTRACT_INTEGRATION_VERSION = make_semver(1, 0, 0)
+
+
+class ClusterAuthRhidpIntegrationParams(PydanticRunParams):
+    pass
+
+
+class ManagedLabelConflictError(Exception):
+    pass
+
+
+OcmApis = dict[str, OCMBaseClient]
+
+
+class ClusterAuthRhidpIntegration(
+    QontractReconcileIntegration[ClusterAuthRhidpIntegrationParams]
+):
+    """Manages the OCM subscription labels for clusters with RHIDP authentication."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+
+        desired = {
+            "subs_labels": self.fetch_desired_state(self.get_clusters(query_func)),
+        }
+        # to figure out wheter to run a PR check of to exit early, a hash value
+        # of the desired state is sufficient
+        return {"hash": DeepHash(desired).get(desired)}
+
+    def get_clusters(self, query_func: Callable) -> list[ClusterV1]:
+        data = cluster_query(query_func)
+        return [
+            c
+            for c in data.clusters or []
+            if integration_is_enabled(self.name, c)
+            # ocm is mandatory
+            and c.ocm
+            and c.spec
+            and c.spec.q_id
+        ]
+
+    def get_environments(self, query_func: Callable) -> list[OCMEnvironment]:
+        return ocm_environment_query(query_func).environments
+
+    def fetch_desired_state(self, clusters: Iterable[ClusterV1]) -> LabelState:
+        """Fetches the desired state of subscription labels for the given clusters."""
+        states: LabelState = {}
+
+        for cluster in clusters:
+            assert cluster.ocm and cluster.spec and cluster.spec.q_id  # make mypy happy
+            label = {}
+            for auth in cluster.auth:
+                if not isinstance(auth, ClusterAuthRHIDPV1) or auth.service != "rhidp":
+                    continue
+                label = {
+                    STATUS_LABEL_KEY: auth.status or StatusValue.ENABLED.value,
+                    AUTH_NAME_LABEL_KEY: auth.name,
+                }
+                if auth.issuer:
+                    label[ISSUER_LABEL_KEY] = auth.issuer
```

Comment:
Any validation on at most one match `auth` in `cluster.auth`, otherwise `label` values may be the last matching one and `ISSUER_LABEL_KEY` can from a different auth.

Maybe use `next` to locate first one to use?

```python
            auth = next(
                (
                    a
                    for a in cluster.auth
                    if isinstance(a, ClusterAuthRHIDPV1) and a.service == "rhidp"
                ),
                None,
            )
```
```

Reply:
Fixed in [471b1d8](https://github.com/app-sre/qontract-reconcile/pull/4338/commits/471b1d8c64604127bf8471d22513e89fc92e0e56)